### PR TITLE
[Re-write] Free fleet adapter using easy-full-control fleet adapter and zenoh bridges

### PR DESCRIPTION
**Pull request by aaronchongth**
## New feature implementation

A large motivation of this re-write, is to fully utilize the well tested tools already available in the ROS ecosystem and Zenoh, and also take advantage of python bindings as much as possible, to allow for easier iterations, testing and contributions.

The use of Zenoh bridges will hopefully also minimize disruption to any existing robot navigation stack, and can be spun up with just a standalone binary with the appropriate configuration file.

The legacy implementation can still be found at the `legacy` branch.

* re-implemented free_fleet using `easy_full_control`, based off the `fleet_adapter_template`, will allow setting up of perform actions, https://github.com/open-rmf/free_fleet/issues/116, https://github.com/open-rmf/free_fleet/issues/141
* fleet adapter to use `zenoh` as the main way of communicating with robots
* change approach to use `zenoh-bridge-ros2dds` as a drop-in method of bridging required ROS 2 topics and actions between robot and fleet adapter
* supports the nav2 stack with the simple `/navigation_to_pose` action, https://github.com/open-rmf/free_fleet/issues/84 
* using `pycdr2` to serialize and deserialize messages, with required message definitions fleshed out, https://github.com/open-rmf/free_fleet/issues/129
* examples using existing simulations from `nav2_bringup`
* updated README, https://github.com/open-rmf/free_fleet/blob/easy-full-control/README.md
* supports Jazzy (other distros not tested yet), https://github.com/open-rmf/free_fleet/issues/138
* linting, https://github.com/open-rmf/free_fleet/issues/100
* support for ROS 1 navigation stack (tested in simulation only, part of integration tests)

## Architecture

This architecture and the use of Zenoh bridges will ideally improve the integration experience, with just setting up the `zenoh-bridge-ros2dds` and its configuration, instead of building and running a separate node.  

This PR does not support the ROS 1 navigation stack, the diagram is just for illustrating the proposed architecture once it is implemented down the road.

## Two robots example using official `nav2_bringup`'s `unique_multi_tb3_simulation`

Although the two robot simulation example is just for testing purposes as it uses a different architecture as intended

## Next TODOs (over subsequent PRs)
* attempt to optimize tf messages (not all are needed)
* map switching support
* docking support
* support for Rolling
* docker images
* releases
* testing and support for other RMW implementations

There are plenty of TODOs left, but this is the bare minimum testable and usable state

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,306 +1,265 @@
...
 
 ```bash
-source ~/ff_ros2_ws/install/setup.bash
-ros2 launch ff_examples_ros2 turtlebot3_world_ff_server.launch.xml
+source ~/ff_ws/install/setup.bash
+ros2 run free_fleet_examples test_navigate_to_pose.py \
+    --frame-id map \
+    --namespace turtlebot3_1 \
+    -x 1.808
+    -y 0.503
```
Comment (xiyuoh) [inline README.md:128]:
```suggestion
ros2 run free_fleet_examples test_navigate_to_pose.py \
    --frame-id map \
    --namespace turtlebot3_1 \
    -x 1.808 \
    -y 0.503
```

Missing backslash, copy and paste doesn't work

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+                elif rep.status == GoalStatus.STATUS_SUCCEEDED:
+                    self.node.get_logger().info(
+                        f'Navigation goal {self.nav_goal_id} reached'
+                    )
+                    return True
+                else:
+                    self.node.get_logger().error(
+                        f'Navigation goal {self.nav_goal_id} status '
+                        f'{rep.status}')
+                    return True
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:149]: In cases where the navigation goal has failed and the robot did not reach the goal waypoint successfully, `_is_navigation_done()` returns True. This might cause RMF to misunderstand that the robot is ready for the subsequent waypoint. It might help to either add a retry mechanism for the same goal waypoint, or trigger a replan.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+            battery_state = SensorMsgs_BatteryState.deserialize(sample.payload)
+            self.battery_soc = battery_state.percentage
+
+        self.battery_state_sub = self.zenoh_session.declare_subscriber(
+            namespacify('battery_state', name),
+            _battery_state_callback
+        )
+
+    def _make_random_goal_id(self):
+        return np.random.randint(0, 255, size=(16)).astype('uint8').tolist()
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:117]: Nit, though its random and a size 16 list, I wonder if there's a more reliable way to prevent overlapping goal IDs.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            f'Commanding [{self.name}] to navigate to {destination.position} '
+            f'on map [{destination.map}]'
+        )
+
+        if destination.map != self.map:
+            self.node.get_logger().error(
+                f'Destination is on map [{destination.map}], while robot '
+                f'[{self.name}] is on map [{self.map}]'
+            )
+            return
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:197]: While this is unlikely, we could trigger a replan instead of returning with nothing to ensure that RMF sends in the next reasonable command. Otherwise the execution object remains waiting to be marked `finished()` without acknowledging that this command is invalid.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+                        f'Navigation goal {nav_goal_id} accepted'
+                    )
+                    self.nav_goal_id = nav_goal_id
+                    return
+
+                self.node.get_logger().error(
+                    f'Navigation goal {nav_goal_id} was rejected'
+                )
+                self.nav_goal_id = None
+                return
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:245]: What happens to `self.execution` in this scenario?

**Review by xiyuoh — CHANGES_REQUESTED** [review]
Thank you for revamping free fleet! Tried out the examples and things are looking fantastic. I was able to bring up both single and multiple TB3 tests by following the instructions on README.

Left some minor comments for the README that are mostly about typos and clarity. For the fleet/robot adapter, I was primarily looking out for edge cases where `self.nav_goal_id` might fall out of sync with `self.execution` and catching any scenarios where the execution object is left unattended if `self.nav_goal_id` is set to None. This may lead to RMF not sending any subsequent commands, or sending outdated nav commands if replans are not being called prompty.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py (hunk 3 above)
Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:117]: I think this should be random enough, it'll be 256^16 combinations, but I'll check out how `nav2` usually does it and use a similar method perhaps

File: README.md [hunk 6]
```diff
@@ -1,306 +1,265 @@
...
 
 ```bash
-source ~/ff_ros1_ws/install/setup.bash
-export TURTLEBOT3_MODEL=burger; roslaunch ff_examples_ros1 turtlebot3_world_ff.launch
-```
+source ~/ff_ws/install/setup.bash
+export ROS_DOMAIN_ID=55
 
-This launch file starts the simulation in `gazebo`, visualization in `rviz`, as well as the simulated navigation stack of the single turtlebot3. Once the simulation and visualization show up, the robot can be commanded as per normal through `rviz` with `2D Nav Goal`.
+ros2 launch free_fleet_examples tb3_simulation_fleet_adapter.launch.xml
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply README.md:146]: Hmmm, I definitely see the merit in that, however I was hoping to get users more familiar with the architecture of RMF, rather than launching and killing everything in one go, with this setup, they can jumble it together if they want to, or they can kill just the fleet adapter, fix something, and launch it again, without killing the entire stack

I'll keep this in mind and monitor the UX for a bit, but I'd probably like to maintain these set of steps for now

File: README.md [hunk 7]
```diff
@@ -1,306 +1,265 @@
...
+Launch the `free_fleet_adapter` with the current example's configurations, verify that `turtlebot3_1` has been added to fleet `turtletbot`.
 
-</br>
+```bash
+source ~/ff_ws/install/setup.bash
+export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp
+export ROS_DOMAIN_ID=55
 
-### Commands and Requests
+ros2 launch free_fleet_examples unique_multi_tb3_simulation_fleet_adapter.launch.xml
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply README.md:209]: replied above

File: README.md [hunk 8]
```diff
@@ -1,306 +1,265 @@
...
-  # maven default-jdk   # Uncomment to install dependencies for message generation
-```
+## Source build and setup
 
-</br>
-
-### Message Generation
-
-Message generation via `FleetMessages.idl` is done using `dds_idlc` from `CycloneDDS`. For convenience, the generated mesasges and files has been done offline and committed into the code base. They can be found [here](./free_fleet/src/messages/FleetMessages.idl).
+Other system depenendencies,
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply README.md]: https://github.com/open-rmf/free_fleet/pull/145/commits/80df6118b8b7c00c9e50dfa2b21b1922df3ef9ec

File: README.md [hunk 9]
```diff
@@ -1,306 +1,265 @@
...
 
-Source ROS 2 and build, 
+For ROS 2, simulations will be launched using the `nav2_bringup` package. Since the `turtlebot3_gazebo` package is not being released past jazzy, users will need to clone the package to access the gazebo models,
 
-```bash
-cd ~/ff_ros2_ws
-source /opt/ros/galactic/setup.bash
-colcon build
+```
+sudo apt install ros-jazzy-nav2-bringup
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply README.md]: We won't need it as `nav2-bringup` will install all that's required by the launch files, but added the update commands, https://github.com/open-rmf/free_fleet/pull/145/commits/80df6118b8b7c00c9e50dfa2b21b1922df3ef9ec

File: README.md [hunk 10]
```diff
@@ -1,306 +1,265 @@
...
-## Installation Instructions
+Supports
+* [Ubuntu 24.04](https://ubuntu.com/blog/ubuntu-desktop-24-04-noble-numbat-deep-dive)
+* [ROS 2 Jazzy](https://docs.ros.org/en/jazzy/index.html)
+* [rmw-cyclonedds-cpp](https://github.com/ros2/rmw_cyclonedds)
+* [Open-RMF on main](https://github.com/open-rmf/rmf)
+* [zenoh-bridge-ros2dds v0.11](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds/releases/tag/0.11.0)
 
-### Prerequisites
+We recommend setting up `zenoh-bridge-ros2dds` with the standalone binaries. After downloading the appropriate released version and platform, extract and use the standalone binaries as is. For source builds of `zenoh-bridge-ros2dds`, please follow the [official guides](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds).
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply README.md:29]: https://github.com/open-rmf/free_fleet/pull/145/commits/80df6118b8b7c00c9e50dfa2b21b1922df3ef9ec

**Comment by aaronchongth** [comment]
Heads up @xiyuoh when you try it out again

While setting up static and integration tests, I've found that setting up a zenoh router between the bridges and the `free_fleet_adapter` provides the best stability and repeatability (IRL and in CI). Folks can reference the [docker compose file](https://github.com/open-rmf/free_fleet/blob/easy-full-control/.github/docker/integration-tests/docker-compose.yaml).

I've updated the documentation, tests and architecture diagrams.

This is not a big change as it was already expected that folks will need to use a zenoh router eventually when setting up everything as a distributed system.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+                    case GoalStatus.STATUS_SUCCEEDED.value:
+                        self.node.get_logger().info(
+                            f'Navigation goal {self.nav_goal_id} reached'
+                        )
+                        return True
+                    case GoalStatus.STATUS_CANCELED.value:
+                        self.node.get_logger().info(
+                            f'Navigation goal {self.nav_goal_id} was cancelled'
+                        )
+                        return True
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:206]: I believe this case is handled under `stop()`, so we can probably leave `CANCELED` scenarios out. When RMF requests for a robot to stop, leading to all goal requests being canceled, `self.execution` and `self.nav_goal_id` are already set to `None` inside the `stop()` and `_handle_stop_navigation()` callbacks respectively. On the other hand, `is_navigation_done()` is only ever called when both `self.execution` and `self.nav_goal_id` are not `None`, so this case would be redundant here.

Suggesting to leave it out because it'd be good to have a single source of truth/logic for goal canceled cases, especially when we are returning `True` and telling RMF that we have successfully completed a navigation task. Feel free to let me know if I'm misunderstanding the usage.

File: README.md [hunk 12]
```diff
@@ -1,309 +1,307 @@
...
-```bash
-sudo apt update && sudo apt install \
-  git wget qtbase5-dev qtchooser qt5-qmake qtbase5-dev-tools \
-  python3-rosdep \
-  python3-vcstool \
-  python3-colcon-common-extensions \
-  # maven default-jdk   # Uncomment to install dependencies for message generation
-```
+> [!WARNING]
+> This has so far only been tested in simulation, and will undergo updates and changes as more testing is performed. Use at your own risk. For the legacy implementation, check out the [`legacy`](https://github.com/open-rmf/free_fleet/tree/legacy) branch.
```
Comment (xiyuoh) [inline README.md:36]: If I'm not wrong you've been testing this on hardware? If so this could be updated 😄

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+        self.battery_soc = 1.0
+
+        self.replan_counts = 0
+
+        self.tf_handler = Nav2TfHandler(
+            self.name,
+            self.zenoh_session,
+            self.tf_buffer,
+            self.node
+        )
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:136]: I assume the intention of using an abstract class `RobotAdapter` is to open up options for integrators who want to tweak the implementation of the Nav2 adapter according to their use case, e.g. adding custom actions in the future.
In that case, some of these parameters can be ported into `RobotAdapter` to ensure that they will be initialized instead of relying on integrators to set them and possibly missing them out. It could also be useful to document this intention inside README to direct integrators to use this abstract class.

This also looks good to brought into the [fleet_adapter_template](https://github.com/open-rmf/fleet_adapter_template/blob/main/fleet_adapter_template/fleet_adapter_template/RobotClientAPI.py) with some tweaks, but that's a goal for another day.

**Review by xiyuoh — CHANGES_REQUESTED** [review]

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py (hunk 13 above)
Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:136]: I would actually leave the setting of member variables to the child classes, in this case `Nav2RobotAdapter`, instead of moving these operations upward to `RobotAdapter`. Keeping `RobotAdapter` as a pure abstract class will help us not limit the constructors of any implemented `RobotAdapter`, but still maintain an overridden API that the final free fleet adapter can call upon

I believe this will help us when we start implementing the `Nav1RobotAdapter`. But in any case, if I don't see any major benefit down the road, I'll follow your advice and help move these upwards into `RobotAdapter` and update `fleet_adapter_template` too

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py (hunk 11 above)
Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:206]: So far the case doesn't do anything other than tell the robot adapter that navigation is done, there was a race condition I encountered while testing this behavior, that was "fixed" with https://github.com/open-rmf/free_fleet/pull/145/commits/9cf376a3fb9064ea43b95135c43443236903fb06#diff-6f7a8ca807eab341270b8ed8d8073dcea66db64ff693dd774d681225f2f526ab

Currently this code block doesn't do anything other than produce a log, and report that navigation is done. The alternative would be more hairy, as it will start going into the logic for replanning.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py (hunk 13 above)
Comment (xiyuoh) [reply free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:136]: Gotcha, I understand where you're coming from. Flexibility is definitely a desirable trait we want in these classes. Regarding member variables, params such as `update_handle`, `fleet_handle` and `node` (and maybe even robot `name`) are pretty  much non-negotiable for any EFC fleet adapter. Having them defined in the abstract class would ensure that users provide them as arguments when setting up their fleet adapters. The alternate case could be users sifting through existing fleet adapter implementations to pick out variables they assume are important, and potentially leaving out these "essentials", making the setup process longer and more confusing. I'm not leaning very strongly to have these added in though, just a suggestion.

If you're reconsidering including them but are still worried about constructor limitations, I recently received a [review comment](https://github.com/open-rmf/fleet_adapter_mir/pull/36#discussion_r1784333490) that suggested wrapping variables in an [`ActionContext`](https://github.com/open-rmf/fleet_adapter_mir/blob/main/fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py#L71) object, and [import it into the abstract class](https://github.com/open-rmf/fleet_adapter_mir/blob/main/fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py#L21-L22) as a single argument. I thought the concept could be useful for this scenario too, allowing us to update the Context object without breaking API if we ever want to add on to the non-nego variables.

With that said, I still struggle to see the need for a pure abstract `RobotAdapter` class that only lives within `free_fleet` when it has nothing specific to free fleet, and instead can be utilized across all types of EFC fleet adapters for different robots. Imo it would be more suitable to port this to the fleet adapter template, and have free fleet adapters inherit the class from there. That way we won't have multiple abstract class across all our fleet adapter repos. If we do want an abstract `RobotAdapter` class under the `free_fleet` repo, it should probably contain some `free_fleet`-specific methods.

Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py]: Ok, there are a few dilemmas here, because the `RobotAdapter` is for testability and consistency over the nav1 and nav2 robot adapters (and any other types of navigation stack we might consider in the future), and not a plugin system,

> potentially leaving out these "essentials"

the essentials can technically be left out if their fleet adapter implementations don't require it (I can't think of why, but it is possible to leave out name and node, and just have the fleet adapter keep track of it)

users might also want to create their own node inside their own `RobotAdapter` perhaps to set up different callback groups for their own uses or optimizations, instead of using the provided node like we usually do

> If you're reconsidering including them but are still worried about constructor limitations

Since the next few items on the list for `free_fleet` is related to implementing the nav1 robot adapter, what exactly can be abstracted into the constructor `RobotAdapter` is yet to be confirmed, and will have a high risk of changing, breaking API. I will look into using `ActionContext`, but I'd say we can leave that for later

> the need for a pure abstract RobotAdapter class that only lives within free_fleet when it has nothing specific to free fleet

The `Nav1RobotAdapter` will be implemented in the same manner such that in the future, the `free_fleet_adapter` is able to control a mixed fleet of Nav1 and Nav2 robots, so I'd say it is rather related to free fleet, although as you said, not specific only to free fleet.

On the other hand `fleet_adapter_template` at the moment is a template, and the final implementation of any fleet adapter based on the template still needs to be written in a separate package, without importing it from `fleet_adapter_template`. Adding an abstracted `RobotAdapter` to `fleet_adapter_template`, although useful as a reference, could lead to even more confusion, as folks can import `RobotAdapter` but implement their own fleet adapter based on the `fleet_adapter_template` at that time . But if any implementation details (member variable name) changes upstream for the imported `RobotAdapter`, without updating their own fleet adapter, things will break. Hence I think it will be clearer to users if they have full visibility of `RobotAdapter`.

I've moved the member variables directly used by the `free_fleet_adapter` into `RobotAdapter`, while keeping everything else API calls (which have been abstracted). I'd say only if we can make the upstream fleet adapter template a library or a plugin system for robot adapters, should we consider moving `RobotAdapter` upstream

**Review by xiyuoh — APPROVED** [review]
After discussing more about the abstract class with Aaron, free fleet seems like a good place to house it since the goal is for a single fleet to integrate with multiple types of robot adapters. It would be extremely useful for future maintainers and community members to contribute their own robot adapters with other nav stacks.

However, since the current iteration of the class is meant for testability and is subjected to change after experimenting with nav1, I suggested it might be more appropriate to flesh out the class methods, and only introduce it after we ensure that there will be no breaking changes (especially with nav1).

Apart from that this PR looks good to me, have tested it out in sim + witnessed the hardware working well with the fleet adapter. Thanks for this massive refactor!

File: free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py [hunk 14]
```diff
@@ -0,0 +1,553 @@
...
+        if map_name != self.map_name:
+            # TODO(ac): test this map related replanning behavior
+            self.replan_counts += 1
+            self.node.get_logger().error(
+                f'Destination is on map [{map_name}], while robot '
+                f'[{self.name}] is on map [{self.map_name}], replan count '
+                f'[{self.replan_counts}]'
+            )
+
+            self._check_update_handle_initialization()
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py:475]: Didn't catch this the previous time, but this check is not required here. If RMF is able to call `navigate()`, it means an [update handle is already successfully created](https://github.com/open-rmf/free_fleet/blob/easy-full-control/free_fleet_adapter/free_fleet_adapter/fleet_adapter.py#L233-L247). Same for nav2 adapter.

File: free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py [hunk 15]
```diff
@@ -0,0 +1,553 @@
...
+    def _check_update_handle_initialization(self):
+        if self.update_handle is None:
+            error_message = \
+                f'Failed to update robot {self.name}, robot adapter has ' \
+                'not yet been initialized with a fleet update handle.'
+            self.node.get_logger().error(error_message)
+            raise RuntimeError(error_message)
+
+    def update(self, state: rmf_easy.RobotState):
+        self._check_update_handle_initialization()
```
Comment (xiyuoh) [inline free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py:442]: This is not required either as the code does not reach here unless there is an update handle. But if you'd like to leave this check in, suggest to print a warning for `_check_update_handle_initialization()` and return early instead of raising RuntimeError as the update handle can still be initialized at the next [update loop](https://github.com/open-rmf/free_fleet/blob/easy-full-control/free_fleet_adapter/free_fleet_adapter/fleet_adapter.py#L221).

File: .github/docker/minimal-nav1-bringup/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,21 @@
+ARG ROS_DISTRO=noetic
+FROM docker.io/ros:$ROS_DISTRO-ros-base
+
+RUN apt update && apt install -y curl ros-$ROS_DISTRO-turtlebot3-navigation ros-$ROS_DISTRO-dwa-local-planner
+
+RUN mkdir -p /tb3 && cd /tb3 \
+  && curl -sL https://github.com/ros-navigation/navigation2/archive/refs/heads/master.tar.gz -o navigation2.tar.gz \
```
Comment (mxgrey) [inline .github/docker/minimal-nav1-bringup/Dockerfile:7]: It looks like this fetches from `master` instead of from [a specific tag](https://github.com/ros-navigation/navigation2/tags). That means the docker file could start failing for users if something is changed upstream. Is that intentional? I can imagine good arguments for this and against this so I don't have a strong opinion on which is better; I just want to make sure the concern was considered.

Comment (mxgrey) [reply .github/docker/minimal-nav1-bringup/Dockerfile:7]: Also if you do want to track the branch instead of a tag you should probably change `master` to `main`. It seems that github is automatically redirecting `master` to refer to `main`, but there's no telling how long they'll keep doing that.

File: free_fleet/free_fleet/ros2_types.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+from dataclasses import dataclass
+from enum import Enum
+
+import pycdr2
+from pycdr2 import IdlStruct
+
+
+# https://github.com/ros2/rcl_interfaces/blob/rolling/builtin_interfaces/msg/Time.msg
+@dataclass
+class Time(IdlStruct, typename='Time'):
```
Comment (mxgrey) [inline free_fleet/free_fleet/ros2_types.py:26]: I'm curious, what's this `typename='...'` syntax inside of the class inheritance brackets? I haven't seen this before, and my Google searches aren't turning up any examples or explanations.

File: free_fleet/free_fleet/utils.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def namespacify(base_name: str, namespace: str, delimiter: str = '/') -> str:
+    """
+    Namespaces a base_name with namespace and delimiter.
+
+    If no namespace is provided, returns the base_name. Otherwise, naively
+    prefixes the base_name with namespace and delimiter, returning the result.
+    """
+    return f'{namespace}{delimiter}{base_name}' if len(namespace) != 0 \
```
Comment (mxgrey) [inline free_fleet/free_fleet/utils.py:32]: I wonder if the condition should be `if not namespace` which I believe would catch both empty string and `None` values.

File: free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py (hunk 15 above)
Comment (mxgrey) [reply free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py:442]: Definitely, let's just return early here if the update handle is not available. That could just be caused by a one-off race condition.

File: free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py [hunk 19]
```diff
@@ -0,0 +1,553 @@
...
+                self.update_handle.more().replan()
+                return False
+
+    def _check_update_handle_initialization(self):
+        if self.update_handle is None:
+            error_message = \
+                f'Failed to update robot {self.name}, robot adapter has ' \
+                'not yet been initialized with a fleet update handle.'
+            self.node.get_logger().error(error_message)
+            raise RuntimeError(error_message)
```
Comment (mxgrey) [inline free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py:439]: In general I would recommend removing this function and checking `update_handle` on a case-by-case basis. For callbacks that are triggered by RMF, there is no need to check `update_handle`. For anything else, throwing an exception is probably too extreme.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py (hunk 11 above)
Comment (mxgrey) [reply free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:205]: Is there a possibility that the `CANCELED` status could be caused by the nav stack itself deciding it cannot reach the destination? E.g. the destination is blocked and therefore the action gets cancelled?

In the event of an unexpected cancellation we should [open an issue ticket](https://github.com/open-rmf/fleet_adapter_mir/blob/adc24ce34ef3b18aa113e995e6f7ba4419939138/fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py#L364) and replan until it succeeds.

File: free_fleet_adapter/launch/fleet_adapter.launch.xml [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+  <arg name="nav_graph_file" description="The graph that this fleet should use for navigation"/>
+  <arg name="server_uri" default="" description="The URI of the api server to transmit state and task information."/>
+  <arg name="output" default="screen"/>
+
+  <!-- Fleet adapter -->
+  <group>
+    <group if="$(var use_sim_time)">
+      <node pkg="free_fleet_adapter"
+            exec="fleet_adapter.py"
+            args="-c $(var config_file) -n $(var nav_graph_file) -s $(var server_uri) -sim"
```
Comment (mxgrey) [inline free_fleet_adapter/launch/fleet_adapter.launch.xml:16]: Instead of needing to branch this launch file logic, it may be worth checking if you can tweak its implementation so that you initialize an `rclpy.Node` early on and then check if the `use_sim_time` parameter is set to true or false rather than having a dedicated command line argument for it.

If there's some reason that can't be done, then you could consider tweaking the argument to so that this line could be `--sim $(use_sim_time)` so you can just feed in the boolean instead of needing to have different branches based on whether or not to include `-sim`.

File: free_fleet_adapter/tests/integration/test_nav1_tf_handler.py [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+            'missing_nav1_tb3', self.zenoh_session, tf_buffer, self.node
+        )
+
+        transform_exists = False
+        for i in range(10):
+            transform = tf_handler.get_transform()
+            if transform is not None:
+                transform_exists = True
+                break
+            time.sleep(1)
```
Comment (mxgrey) [inline free_fleet_adapter/tests/integration/test_nav1_tf_handler.py:54]: Is this integration test meant to be run with real hardware in the loop? If not, I'm surprised we'd have a sleep of one entire second in between pinging for transforms. I would think ~10ms should be enough wait if this is just testing against software.

File: free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py (hunk 11 above)
Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py:213]: Yup thanks for catching that, I should have referred to the mir fleet adapter to see what was missing, I've added it in https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py (hunk 14 above)
Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py]: removed in https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py (hunk 15 above)
Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py]: Removed and returning early in https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: .github/docker/minimal-nav1-bringup/Dockerfile (hunk 16 above)
Comment (aaronchongth) [reply .github/docker/minimal-nav1-bringup/Dockerfile]: Thanks for catching that, that must have been a copy-paste mistake. It definitely makes sense to track a specific tag, since all we need is the map scan. Change to use a tag in https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: free_fleet/free_fleet/ros2_types.py (hunk 17 above)
Comment (aaronchongth) [reply free_fleet/free_fleet/ros2_types.py]: Good point, I think it might have been a legacy thing, or something else. I followed the [bridge examples](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds/blob/examples/examples/python/src/fibonnacci_action_client.py) in the past and forgot they were there. Removed in https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: free_fleet/free_fleet/utils.py (hunk 18 above)
Comment (aaronchongth) [reply free_fleet/free_fleet/utils.py]: Yeah that'll catch more issues for sure, https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py (hunk 19 above)
Comment (aaronchongth) [reply free_fleet_adapter/free_fleet_adapter/nav1_robot_adapter.py]: Removed and returned early in https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: free_fleet_adapter/launch/fleet_adapter.launch.xml (hunk 20 above)
Comment (aaronchongth) [reply free_fleet_adapter/launch/fleet_adapter.launch.xml]: That's a good point, I ended up with some conditional variables in the xml instead, https://github.com/open-rmf/free_fleet/pull/145/commits/cccb4033c024b5035a97c7cda146f70970b24baf

File: free_fleet_adapter/tests/integration/test_nav1_tf_handler.py (hunk 21 above)
Comment (aaronchongth) [reply free_fleet_adapter/tests/integration/test_nav1_tf_handler.py]: It was meant to give some time for initialization, as my initial tests in the past (without a zenoh router) too a while for peers to find each other and transforms to come through. Now that we encourage a zenoh router in the integration, it definitely does not warrant such a long sleep now. I've cleaned up all the tests a little bit, so it just waits for 2 seconds before checking. 1 second was not enough in my tests :sweat_smile: 

changes made in https://github.com/open-rmf/free_fleet/pull/145/commits/e3a5a1f10e6250ed34a1dfee816937f029e42d13

**Review by mxgrey — APPROVED** [review]
-20k net lines of code while also improving capabilities is such a rare treat
